tests/client/api_restful/test_send_message.c

```diff
@@ -175,24 +184,129 @@ void test_send_core_message_tagged_data_tag_max_len() {
+                              0x32, 0x43, 0x56, 0xB6, 0x53, 0x65, 0xD0, 0xE3, 0x9A, 0x30, 0x3A, 0xC5, 0xBB,
+                              0xAE, 0x83, 0xD0, 0x13, 0xBC, 0x76, 0xB6, 0xC5, 0xE6, 0xFD, 0xCD, 0x2E, 0x72,
+                              0xEC, 0x80, 0x41, 0x33, 0xF3, 0x7B, 0xC0, 0x3E, 0x2A, 0x35, 0xC4, 0xA9};
+
+  // get address from mnemonic seed
+  address_t addr_send, addr_recv;
+  ed25519_keypair_t sender_key = {};
+  char bech32_sender[65] = {};
+  char bech32_receiver[65] = {};
+  // IOTA BIP44 Paths: m/44'/4128'/Account'/Change'/Index'
+  TEST_ASSERT(address_keypair_from_path(mnemonic_seed, sizeof(mnemonic_seed), "m/44'/4218'/0'/0'/0'", &sender_key) ==
+              0);
+  TEST_ASSERT(ed25519_address_from_path(mnemonic_seed, sizeof(mnemonic_seed), "m/44'/4218'/0'/0'/0'", &addr_send) == 0);
+  TEST_ASSERT(ed25519_address_from_path(mnemonic_seed, sizeof(mnemonic_seed), "m/44'/4218'/0'/0'/1'", &addr_recv) == 0);
+
+  char const* const hrp = "atoi";
+  address_to_bech32(&addr_send, hrp, bech32_sender, sizeof(bech32_sender));
+  address_to_bech32(&addr_recv, hrp, bech32_receiver, sizeof(bech32_receiver));
+  printf("sender: %s\nreceiver: %s\n", bech32_sender, bech32_receiver);
+
+  transaction_payload_t* tx = tx_payload_new();
+  TEST_ASSERT_NOT_NULL(tx);
+
+  // get outputs from an address
+  uint64_t send_amount = 1000000;  // send out 1Mi
+  res_outputs_id_t* res = res_outputs_new();
+  outputs_query_list_t* query_param = outputs_query_list_new();
+  outputs_query_list_add(&query_param, QUERY_PARAM_ADDRESS, bech32_sender);
+  TEST_ASSERT(get_outputs_id(&ctx, query_param, res) == 0);
+  TEST_ASSERT(res_outputs_output_id_count(res) > 0);
+
+  // dump outputs for debugging
+  // for (size_t i = 0; i < res_outputs_output_id_count(res); i++) {
+  //   printf("output[%zu]: %s\n", i, res_outputs_output_id(res, i));
+  // }
+
+  // fetch output data from output IDs
+  uint64_t total_balance = 0;
+  for (size_t i = 0; i < res_outputs_output_id_count(res); i++) {
+    res_output_t* output_res = get_output_response_new();
+    printf("fetch output: %s\n", res_outputs_output_id(res, i));
+    TEST_ASSERT(get_output(&ctx, res_outputs_output_id(res, i), output_res) == 0);
+    if (!output_res->is_error) {
+      if (output_res->u.data->output->output_type == OUTPUT_EXTENDED) {
+        output_extended_t* o = (output_extended_t*)output_res->u.data->output->output;
+        total_balance += o->amount;
+        // check balane
+        if (total_balance >= send_amount) {
+          // add it to tx payload and break
+          tx_payload_add_input(tx, 0, output_res->u.data->tx_id, output_res->u.data->output_index, &sender_key);
+          get_output_response_free(output_res);
+          break;
+        } else {
+          // add it to tx payload and fetch next output data
+          tx_payload_add_input(tx, 0, output_res->u.data->tx_id, output_res->u.data->output_index, &sender_key);
```

Comment:
`tx_payload_add_input(tx, 0, output_res->u.data->tx_id, output_res->u.data->output_index, &sender_key);` can be extracted from `if `and `else` blocks and moved above them.

Reply:
nice catch :)